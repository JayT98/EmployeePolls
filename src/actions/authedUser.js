export const LOGOUT_AUTHED_USER = "LOGOUT_AUTHED_USER";
export const SET_AUTHED_USER = "SET_AUTHED_USER";

export function setAuthedUser(authedUser){
    return {
        type: SET_AUTHED_USER,
        authedUser
    }
}

export function logoutAuthedUser() {
    return {
        type: LOGOUT_AUTHED_USER
    }
}

export function handleLogin(username, password){
    return(dispatch, getState) => {
        const {users} = getState();

        const user = Object.values(users).find((user) =>
            user.id === username && user.password === password
        );

        if(!!user){
            return dispatch(setAuthedUser(user));
        }
    };
}

export function handleLogout() {
    return (dispatch) => {
        return dispatch(logoutAuthedUser());
    }
}
