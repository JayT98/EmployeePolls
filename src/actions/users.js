export const ADD_QUESTION_USER = "ADD_QUESTION_USER";
export const ADD_ANSWER_USER = "ADD_ANSWER_USER";
export const RECEIVE_USERS = "RECEIVE_USERS";

export function receiveUsers(users) {
    return {
        type: RECEIVE_USERS,
        users
    }
}

export function addQuestionUser({author, id}) {
    return {
        type: ADD_QUESTION_USER,
        author,
        qid: id,
    };
}

export function addAnswerUser(authedUser, qid, answer) {
    return {
        type: ADD_ANSWER_USER,
        authedUser,
        qid,
        answer
    };
}
