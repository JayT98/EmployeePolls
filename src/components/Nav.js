import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {handleLogout} from '../actions/authedUser';

const Nav = ({dispatch, authedUserId}) => {
    const logout = (e) => {
        e.preventDefault();
        dispatch(handleLogout());
    };

    return (
        <nav className='w-full h-14 flex justify-between border-b-2 border-slate-600/200'>
            <div className='flex'>
            <Link to="/"
                    className='
                        font-bold w-24 h-14
                        text-center
                        flex justify-center items-center
                        hover:border-b-2 border-black'>Home
            </Link>
            <Link to="/leaderboard"
                    className='
                        font-bold w-32 h-14
                        text-center
                        flex justify-center items-center
                        hover:border-b-2 border-black'>Leaderboard
            </Link>
            <Link to="/new"
                    className='
                        font-bold w-24 h-14
                        text-center
                        flex justify-center items-center
                        hover:border-b-2 border-black'>New
            </Link>
            </div>
            <div className=' w-64 flex justify-between'>
            <span
                className='font-bold h-14 text-slate-700 flex items-center'
                data-testid= 'user-information'>
                User: {authedUserId}
            </span>
            <button
                onClick={logout}
                className='  font-medium w-24 h-14
                            text-center
                            hover:border-b-2 border-black'
            >
                Logout
            </button>
            </div>
        </nav>
    );
};

const mapStateToProps = ({authedUser}) => ({
    authedUserId: authedUser.id,
}) 

export default connect(mapStateToProps)(Nav);