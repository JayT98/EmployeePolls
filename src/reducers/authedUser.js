import {LOGOUT_AUTHED_USER, SET_AUTHED_USER} from '../actions/authedUser';

export default function authedUser(state = null, action) {
    switch (action.type) {
        case SET_AUTHED_USER:
            return action.authedUser;
        case LOGOUT_AUTHED_USER:
            return null;
        default:
            return state;
    }
}