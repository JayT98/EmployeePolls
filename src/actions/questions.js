import {saveQuestion, saveQuestionAnswer} from '../data/api';
import {addAnswerUser, addQuestionUser} from './users';

export const RECEIVE_QUESTIONS = 'RECEIVE_QUESTIONS';
export const ADD_QUESTION = 'ADD_QUESTION';
export const ADD_ANSWER_QUESTION = 'ADD_ANSWER_QUESTION';

export function receiveQuestions(questions) {
    return {
        type: RECEIVE_QUESTIONS,
        questions
    };
}

function addQuestion(question){
    return {
        type: ADD_QUESTION,
        question
    };
}

export function addAnswerQuestion(author, qid, answer) {
    return {
        type: ADD_ANSWER_QUESTION,
        author,
        qid,
        answer
    };
}

export function handleAddQuestion(firstOption, secondOption) {
    return (dispatch, getState) => {
        const {authedUser} = getState();

        return saveQuestion(firstOption, secondOption, authedUser)
                .then((question) => {
                    dispatch(addQuestion(question));
                    dispatch(addQuestionUser(question))
                });
    };
}

export function handleAddAnswer(questionId, answer) {
    return (dispatch, getState) => {
        const {authedUser} = getState();

        return saveQuestionAnswer(authedUser.id, questionId, answer)
                .then(() => {
                    dispatch(addAnswerQuestion(authedUser.id, questionId, answer));
                    dispatch(addAnswerUser(authedUser.id, questionId, answer))
                });
    };
}