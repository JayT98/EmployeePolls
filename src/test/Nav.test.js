import {render} from "@testing-library/react";
import {Provider} from "react-redux";
import {store} from "../store";
import {BrowserRouter} from "react-router-dom";
import React from "react";
import Nav from "../components/Nav";
import {setAuthedUser} from "../actions/authedUser";

describe("Nav", () => {
    it("should render the component", () => {
        store.dispatch(setAuthedUser({id: "sarahedo", password: ""}));

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Nav/>
                </BrowserRouter>
            </Provider>
        );
        expect(component).toBeDefined();
        expect(component).toMatchSnapshot();
    });

    it("should display username of logged in user", () => {
        store.dispatch(setAuthedUser({id: "sarahedo", password: ""}));

        const component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Nav/>
                </BrowserRouter>
            </Provider>
        );

        const userSpanElement = component.getByTestId("user-information");
        expect(userSpanElement.textContent).toBe("User: sarahedo");

    });
});
