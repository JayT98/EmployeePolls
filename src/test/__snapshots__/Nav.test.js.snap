// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="w-full h-14 flex justify-between border-b-2 border-slate-600/200"
      >
        <div
          class="flex"
        >
          <a
            class=" font-bold w-24 h-14 text-center flex justify-center items-center hover:border-b-2 border-black"
            href="/"
          >
            Home
          </a>
          <a
            class=" font-bold w-32 h-14 text-center flex justify-center items-center hover:border-b-2 border-black"
            href="/leaderboard"
          >
            Leaderboard
          </a>
          <a
            class=" font-bold w-24 h-14 text-center flex justify-center items-center hover:border-b-2 border-black"
            href="/new"
          >
            New
          </a>
        </div>
        <div
          class=" w-64 flex justify-between"
        >
          <span
            class="font-bold h-14 text-slate-700 flex items-center"
            data-testid="user-information"
          >
            User: 
            sarahedo
          </span>
          <button
            class="  font-medium w-24 h-14 text-center hover:border-b-2 border-black"
          >
            Logout
          </button>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="w-full h-14 flex justify-between border-b-2 border-slate-600/200"
    >
      <div
        class="flex"
      >
        <a
          class=" font-bold w-24 h-14 text-center flex justify-center items-center hover:border-b-2 border-black"
          href="/"
        >
          Home
        </a>
        <a
          class=" font-bold w-32 h-14 text-center flex justify-center items-center hover:border-b-2 border-black"
          href="/leaderboard"
        >
          Leaderboard
        </a>
        <a
          class=" font-bold w-24 h-14 text-center flex justify-center items-center hover:border-b-2 border-black"
          href="/new"
        >
          New
        </a>
      </div>
      <div
        class=" w-64 flex justify-between"
      >
        <span
          class="font-bold h-14 text-slate-700 flex items-center"
          data-testid="user-information"
        >
          User: 
          sarahedo
        </span>
        <button
          class="  font-medium w-24 h-14 text-center hover:border-b-2 border-black"
        >
          Logout
        </button>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
